### PR TITLE
Sneak: Strip down version

### PR DESCRIPTION
**Pull request by SmallJoker**
This pull does some changes in the new sneak code:
- Variables `m_old_node_below*` and `m_need_to_get_new_sneak_node` are deprecated now
- ~Remove variable `got_teleported`~ already merged
- Remove `air` node check for expired sneak nodes
- Various code style corrections (ignore whitespaces to review)
- Compress `detectSneakLadder` and `detectSneakLedge` into four entire code lines
- Makes sneak ladders feel like climbing
- Disallow and reallow double jumps (two nodes at once) - please decide what you want
- Comparison video: https://www.youtube.com/watch?v=coTcDzEfku4 (already outdated)

Features:
- [x] ~No damage evading on sneak ladders~ already merged
- [x] ~No jitter when moving against sneak limit (#5449)~ already merged
- [ ] No weird sinking
- [x] slower 1 row sneak ladder (using #5527)
- [x] ~slower 2 row sneak ladders~ almost the same speed as in the old sneak code
- [x] No weird behaviour on [stairs](https://i.imgur.com/EtkNVPZ.png)
- [x] 8 times fewer direction checks for sneak ladders + ledges **\*1**
- [x] Jumping on node edges when sneaking (required to make sneak ladders work properly)
- [x] Walking around ledge edges, like [here](https://camo.githubusercontent.com/c132b89c7c9b7480ef84c6506a6238b01e0e586f/687474703a2f2f6936332e74696e797069632e636f6d2f786466746f6c2e6a7067) is possible (see also the video above)

**\*1**) With this pull it will only check the nodes above the sneak node, which incredibly reduces the total checks for sneak ladders & ledges down to _2_, previously up to _16_ for both **!** 

What has not changed: It is still possible to deactivate the sneak glitch or sneak itself entirely using `set_physics_override`.

### CODE REVIEW

**Comment by Fixer-007** [comment]
What I see:
+ sneak ladders are indeed slower, with speeds close to using wooden ladders
+ double jump is possible
+ no jitter at the limit of the node
+ sneaking on door is weird
+ tried some damage evasion - seen nothing new apart from https://github.com/minetest/minetest/issues/329
+ teleport to center on stairs/doors still present
+ one way sneaking + instamove on stairs still present

**Comment by SmallJoker** [comment]
> sneaking on door is weird

```Lua
	-- collision_box needed otherise selection box would be full node size
	collision_box = {
		type = "fixed",
		fixed = {-15/32, 13/32, -15/32, -13/32, 1/2, -13/32},
	},
```
[Doors Saucecode](https://github.com/minetest/minetest_game/blob/master/mods/doors/init.lua#L98-L102)
The new sneak code ensures that the player won't fall off when releasing the sneak key. The doors use a very minimalistic collision box, where the player is now centered on.

**Comment by paramat** [comment]
```
+				// Detect sneak ladder:
+				// Node two meters above sneak node must be solid
+				node = map->getNodeNoEx(m_sneak_node + v3s16(0,2,0),
+					&is_valid_position);
+				m_sneak_ladder_detected = is_valid_position &&
+					nodemgr->get(node).walkable;
```
Might be too simple and triggered by non-sneak ladders.
To make climb reasonable, the node 3 nodes above sneak node must not be walkable, otherwise there is no ledge to grab on to.

**Comment by paramat** [comment]
> Reduces climbing speed of sneak ladders !
1-row sneak ladders are slower than using regular ladders

The idea of new sneak code is to replicate old sneak ladder / jump behaviour, so climb speed must be similar to before and be unreasonably fast.

**Comment by SmallJoker** [comment]
@paramat I believe that the players would rather accept the sneak ladders if they weren't somewhat fast like now. It is not clear what the actual targets are. Is sneak now going to be fixed, changed or kept the same way? To decide this we need more player feedbacks about their idea of the sneak behaviour.

Damage evading, even on sneak ladders, is no longer possible. Some may disagree with this change too since it breaks the old, cheating behaviour.
If there are Windows users who would like to test it: [32-Bit binary,  d5aa9a0](https://www.dropbox.com/s/oco1df1q68xukkw/minetest_sneak_5533_170408_c28a843_3.7z) ([msvcr100.dll error fix](https://www.microsoft.com/de-ch/download/details.aspx?id=5555))

**Comment by paramat** [comment]
It's clear from the discussions that the players that like old sneak generally want sneak ladders to be as fast as they were before, because they like the mobility of the old code. They want the new code to have an option (now switched using 'sneak glitch') that replicates sneak ladder and sneak jump as closely as possible. The discussions have made this very clear and i think there has been more than enough player feedback over the last few weeks.
Are the slower sneak ladders in your code unavoidable?

**Comment by Fixer-007** [comment]
Damage evasion fixes? Very interesting, will try.

File: src/localplayer.cpp [hunk 1]
```diff
@@ -95,62 +91,21 @@ LocalPlayer::~LocalPlayer()
...
-			b_max = box;
-		else if (box.MaxEdge.Y == b_max.MaxEdge.Y)
-			b_max.addInternalBox(box);
-	}
-	return aabb3f(v3f(b_max.MinEdge.X, b_max.MaxEdge.Y, b_max.MinEdge.Z), b_max.MaxEdge);
-}
 
-#define GETNODE(map, p3, v2, y, valid) \
-	(map)->getNodeNoEx((p3) + v3s16((v2).X, y, (v2).Y), valid)
+	if (nodeboxes.size() == 0) {
```
Comment (sfan5) [inline src/localplayer.cpp]: `getTopBoundingBox` returning only the maximum extent of the **topmost** bounding box instead of a random collection of all bboxes together is a feature not a bug :-1:

File: src/localplayer.cpp [hunk 2]
```diff
@@ -319,46 +235,43 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
+		for (s16 z = -1; z <= 1; z++) {
+			const v3s16 p = current_node + v3s16(x,0,z);
+			const v3f pf = intToFloat(p, BS);
+			const v2f diff(position.X - pf.X, position.Z - pf.Z);
+			f32 distance_f = diff.getLength();
 
 			if (distance_f > min_distance_f ||
-					fabs(player_p2df.X-node_p2df.X) > (.5+.1)*BS + sneak_max.X ||
-					fabs(player_p2df.Y-node_p2df.Y) > (.5+.1)*BS + sneak_max.Z)
+					fabs(diff.X) > 0.6 * BS + sneak_max.X ||
```
Comment (sfan5) [inline src/localplayer.cpp]: keep the `(.5+.1)*BS`

File: src/localplayer.cpp (hunk 1 above)
Comment (SmallJoker) [inline src/localplayer.cpp]: Using the old code it's no longer possible to sneak on the most outer side of stairs or similar formed collision boxes that aren't flat since the movement is limited to the top box even if there are more collision boxes below.
However, will rename this function.

File: src/localplayer.cpp (hunk 2 above)
Comment (SmallJoker) [inline src/localplayer.cpp]: Where does this magic number come from at all? I would like to document this properly.

File: src/localplayer.cpp [hunk 3]
```diff
@@ -388,31 +301,79 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
+						!nodemgr->get(node).walkable;
+				}
+			}
 		}
 	}
 
+	if (m_sneak_ladder_detected) {
+		// restore legacy behaviour (this makes the m_speed.Y hack necessary)
+		sneak_max = v3f(0.4 * BS, 0, 0.4 * BS);
+	}
```
Comment (sfan5) [inline src/localplayer.cpp]: It's very confusing to move this code far away from the definition of `sneak_max`

File: src/localplayer.cpp [hunk 4]
```diff
@@ -388,31 +301,79 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
+
+	/*
+		If sneaking, keep in range from the last walked node and don't
+		fall off from it
```
Comment (sfan5) [inline src/localplayer.cpp]: <i>If sneaking, keep on top of last walked node and don't fall off</i>

File: src/localplayer.cpp [hunk 5]
```diff
@@ -388,31 +301,79 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
+
+		position.X = rangelim(position.X,
+			sn_f.X + bmin.X - sneak_max.X, sn_f.X + bmax.X + sneak_max.X);
+		position.Z = rangelim(position.Z,
+			sn_f.Z + bmin.Z - sneak_max.Z, sn_f.Z + bmax.Z + sneak_max.Z);
+
+		if (position.X != old_pos.X)
+			m_speed.X = 0;
+		if (position.Z != old_pos.Z)
+			m_speed.Z = 0;
```
Comment (sfan5) [inline src/localplayer.cpp]: Which problems does this fix?

Comment (SmallJoker) [inline src/localplayer.cpp]: Jitter when walking against the limits

File: src/localplayer.cpp [hunk 6]
```diff
@@ -388,31 +301,79 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
+
+		if (m_speed.Y < 0) {
+			position.Y = MYMAX(position.Y, sn_f.Y + bmax.Y);
+			if (position.Y > old_pos.Y) {
+				CollisionInfo sn_info;
```
Comment (sfan5) [inline src/localplayer.cpp]: what is this / why is it necessary?
why isn't it necessary for `m_speed.X` or `m_speed.Z`?

File: src/localplayer.cpp [hunk 7]
```diff
@@ -388,31 +301,79 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
+				sn_info.old_speed = v3f(m_speed);
+				m_speed.Y = 0;
+				sn_info.new_speed = v3f(m_speed);
+
+				collision_info->push_back(sn_info);
+			}
+		}
+
+		if (fabs(m_speed.Y) < 0.01f)
+			touching_ground = true;
```
Comment (sfan5) [inline src/localplayer.cpp]: does this reintroduce any bugs? How is the `0.01` determined?

File: src/localplayer.h [hunk 8]
```diff
@@ -122,9 +121,14 @@ class LocalPlayer : public Player
 
 	f32 getPitch() const { return m_pitch; }
 
-	void setPosition(const v3f &position)
+	void setPosition(const v3f &position, bool clear_sneak = true)
```
Comment (sfan5) [inline src/localplayer.h]: IMO the old way this was done is better because it has less overhead

File: src/localplayer.cpp (hunk 6 above)
Comment (SmallJoker) [inline src/localplayer.cpp]: > // Handle only fall damage
// (because otherwise walking against something in fast_move kills you)
[clientenvironment.cpp](https://github.com/minetest/minetest/blob/master/src/clientenvironment.cpp#L221)
Will add the collision entry for X,Z axis too.

File: src/localplayer.cpp [hunk 9]
```diff
@@ -388,31 +301,79 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
+	if (m_sneak_ladder_detected) {
+		// restore legacy behaviour (this makes the m_speed.Y hack necessary)
+		sneak_max = v3f(0.4 * BS, 0, 0.4 * BS);
+	}
+
+	/*
+		If sneaking, keep in range from the last walked node and don't
+		fall off from it
+	*/
+	if (could_sneak && m_sneak_node_exists) {
```
Comment (sfan5) [inline src/localplayer.cpp]: Does moving this after the *find sneak node* code cause any problems?

File: src/localplayer.cpp (hunk 7 above)
Comment (SmallJoker) [inline src/localplayer.cpp]: Originally supposed to eliminate floating point errors. But the collision calculation sets this one properly to zero when colliding. Will replace with `0`.

File: src/localplayer.cpp [hunk 10]
```diff
@@ -95,62 +91,21 @@ LocalPlayer::~LocalPlayer()
...
-			b_max.addInternalBox(box);
-	}
-	return aabb3f(v3f(b_max.MinEdge.X, b_max.MaxEdge.Y, b_max.MinEdge.Z), b_max.MaxEdge);
-}
 
-#define GETNODE(map, p3, v2, y, valid) \
-	(map)->getNodeNoEx((p3) + v3s16((v2).X, y, (v2).Y), valid)
+	if (nodeboxes.size() == 0) {
+		b_max.reset(-BS, -BS, -BS);
+		return b_max;
```
Comment (sfan5) [inline src/localplayer.cpp]: this case doesn't need to be handled as this function is not called on non-walkable nodes

Comment (SmallJoker) [inline src/localplayer.cpp]: Edge case for nodeboxes with no nodebox or collision box.

File: src/localplayer.cpp [hunk 11]
```diff
@@ -95,62 +91,21 @@ LocalPlayer::~LocalPlayer()
...
-			node = GETNODE(map, pos, vec, 2, &is_valid_position);
-			ok = is_valid_position && w == nodemgr->get(node).walkable;
-		}
+	std::vector<aabb3f>::const_iterator it = nodeboxes.begin();
+	b_max = aabb3f(it->MinEdge, it->MaxEdge);
 
-		if (ok)
-			return true;
-	}
+	for (++it; it != nodeboxes.end(); ++it)
```
Comment (sfan5) [inline src/localplayer.cpp]: putting the first `++it` into this `for` loop is a little confusing

File: src/localplayer.cpp (hunk 1 above)
Comment (sfan5) [inline src/localplayer.cpp]: This is intended, I did not want players to *fly* above the stairs when sneaking

File: src/localplayer.cpp (hunk 2 above)
Comment (sfan5) [inline src/localplayer.cpp]: `.5` is the distance node center -> node edge
`.1` dunno

File: src/localplayer.cpp (hunk 10 above)
Comment (sfan5) [inline src/localplayer.cpp]: well then return `aabb3f(0, 0, 0)`

File: src/localplayer.cpp [hunk 12]
```diff
@@ -388,31 +301,79 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
+			sn_f.X + bmin.X - sneak_max.X, sn_f.X + bmax.X + sneak_max.X);
+		position.Z = rangelim(position.Z,
+			sn_f.Z + bmin.Z - sneak_max.Z, sn_f.Z + bmax.Z + sneak_max.Z);
+
+		if (position.X != old_pos.X)
+			m_speed.X = 0;
+		if (position.Z != old_pos.Z)
+			m_speed.Z = 0;
+
+		if (m_speed.Y < 0) {
```
Comment (sfan5) [inline src/localplayer.cpp]: also why this condition?

File: src/localplayer.cpp (hunk 6 above)
Comment (sfan5) [inline src/localplayer.cpp]: hm why does this code work without this then?

File: src/localplayer.cpp (hunk 9 above)
Comment (SmallJoker) [inline src/localplayer.cpp]: Could also be moved _above_ the "find sneak node" code, yes. It must be under the collision detection to ensure the speed is handled in last instance from the sneak code.

File: src/localplayer.cpp (hunk 1 above)
Comment (SmallJoker) [inline src/localplayer.cpp]: The problem caused by this change is the weird unidirectional movement on stairs: https://i.imgur.com/EtkNVPZ.png

File: src/localplayer.cpp (hunk 2 above)
Comment (SmallJoker) [inline src/localplayer.cpp]: hehe :)
Okay, will revert this change.

File: src/localplayer.cpp (hunk 1 above)
Comment (sfan5) [inline src/localplayer.cpp]: Yeah that's very unfortunate indeed, however I think allowing people to "fly" on the outside side of stairs is not a good solution.

File: src/localplayer.cpp (hunk 9 above)
Comment (sfan5) [inline src/localplayer.cpp]: Well why did you move it then? o.O

File: src/localplayer.cpp (hunk 12 above)
Comment (SmallJoker) [inline src/localplayer.cpp]: Once upon a time while testing this caused some weird half-sneaking on the side of a node.
Problem seems to be gone with the changes afterwards. Removing.

File: src/localplayer.cpp [hunk 13]
```diff
@@ -319,46 +234,94 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
...
+			// Move player on top of the node when falling
+			// The m_speed.Y < 0 check reduces the climb speed on sneak ladders
+			position.Y = MYMAX(position.Y, sn_f.Y + bmax.Y);
+			if (position.Y > old_pos.Y)
+				m_speed.Y = 0;
+		}
+
+		if (m_speed.Y == 0)
+			touching_ground = true;
```
Comment (SmallJoker) [inline src/localplayer.cpp]: Note for the reviewers:
This line allows jumping on node edges while sneaking. Without it, the player would be stuck as soon the sneak ladder isn't detected anymore -> when reaching its top.

**Comment by Fixer-007** [comment]
Damage evasion like I described here https://github.com/minetest/minetest/issues/329#issuecomment-289164689 and https://github.com/minetest/minetest/issues/329#issuecomment-289311843 is still possible.

**Comment by SmallJoker** [comment]
1-row sneak ladder is now possible since #5527 - also trimmed this one to a minimum.
Did quite a few tests with it and it all works as expected. Am I missing something?

**Comment by paramat** [comment]
Will test.
I do feel the original high sneak ladder speed will generally be wanted, it's good to see this is possible if wanted.

**Comment by azekillDIABLO** [comment]
this should really be merged :+1:

**Comment by paramat** [comment]
When testing this all the issues mentioned throughout #5438 should be tested.

**Comment by paramat** [comment]
Maybe leave docs until last to avoid conflicts?

**Comment by paramat** [comment]
Testing.

Sneak-jump:
Has quite a large and jolting teleport effect near the top of the jump similar to my first attempt at sneak-jump. If you release sneak and jump keys just before the teleport you end up hovering in this position unable to jump onto the ledge;

screenshot_20170413_161300

and then drop when using any move key.

1-column, 2-column and corner sneak ladders:
Is now a series of individual jumps instead of a smooth motion, so is slower than classic code. This repeated jumping doesn't feel good and will be irritating on long climbs.
Control on descent (holding and occasionally releasing sneak) is not as smooth and fine as current code, there is a 'bounce effect'.

It's now possible to sneak across the top of a stair node as if it were a full node, you can sneak above the 'missing part' of the stair node. This is not a big problem but is less realistic than current code.

> falling down if the node forward down is walkable but a nodebox.

^ Still present but not a problem.

> You can't walk across a default:snow or a 1/16 node while holding shift. To see this: drop a single node of default:snow on a flat surface, stand on the edge, hold shift and try to walk over it. You walk into an invisible wall and can't cross with shift held.

^ Still happens but invisible barrier is smaller, just like classic sneak code.

Jumping when at the limit of sneak works but jumping forwards has a much reduced forward speed so is not very useful.

@SmallJoker How do i enable faster climb on sneak ladders?

**Comment by paramat** [comment]
> Reduces climbing speed of sneak ladders slightly

Is an understatement. Players will not accept this slow series of jumps up a sneak ladder. We need to have a fast smooth climb as in current code.

**Comment by paramat** [comment]
With this PR i can't sneak on 1/32 slabs either above other nodes or not, but i can in current code.

**Comment by paramat** [comment]
Yes testing latest version with 3 commits.

**Comment by paramat** [comment]
> The idea behind this behaviour is basically:
    Higher attractivity of ladders and alternative transport
    Easier to tolerate for server admins (seriously, the current sneak ladder is just a fast elevator with damage-protection)
    Bonus: Closer to reality. Our super-Sam doesn't want to stay between two steps either

Yes the idea is fine in isolation, but we are trying to keep players happy who desperately want an option for behaviour as close as possible to classic sneak. Also i disagree with 'normalising' climbing made possible by holes in a wall 2m apart, it's too easy, if it's normalised it may be moved into default behaviour.

**Comment by paramat** [comment]
Testing with Line 266 changed for faster climb.
Simple sneak ladder is fast again but a series of abrupt teleports insted of a smooth motion, much faster than classic code now at 10n/s.
2-column and corner sneak ladders are the same but twice as fast, too fast now, about 20 nodes per second.
:-1: For current PR with either slow or fast sneak ladders, sorry.
However other advantages of your code may be good to have if you can add them to current sneak-jump and sneak-ladder behavioour.

File: src/localplayer.cpp (hunk 9 above)
Comment (SmallJoker) [inline src/localplayer.cpp]: Because otherwise it would be above the collision check, which causes the weird sinking.

**Comment by SmallJoker** [comment]
Measured the speed on a sneak ladder, 90 nodes upwards:
Custom fly speed: 4.59s (19.61 n/s) (`movement_speed_fast = 20`)
Old: 14.41s (6.25 n/s)
New: 12.76s (7.05 n/s), faster than original!
PR: 17.91s (5.02 n/s), still 2 n/s faster than climbing a regular ladder.
That's 20% slower than the original speed.
Test code: https://pastebin.com/MtRVGxYx

**Comment by paramat** [comment]
I think the speed depends on the type of ladder, with the single column ladder being slowest. Is this speed for the double-column ladders?

**Comment by paramat** [comment]
5 n/s seems ok but i still dislike the lack of smooth continuous movement and there are other issues.

**Comment by SmallJoker** [comment]
Measurements of 1-row sneak ladders (+/- compared with 2-row sneak ladders)
Old code: 12.71s (7.08 n/s), +0.8 n/s
Current: 13.32s (6.76 n/s), -0.3 n/s
This PR: 17.88s (5.03 n/s), +0 n/s

Well, at least it's consistent with 2-row ones :/

**Comment by paramat** [comment]
Thanks, perhaps i was thinking of when i edited line 266, which made 2-column ladders faster than 1-column.
7 to 5 is more significant than 6 to 5. What's nice about current code is that all ladders are equal in speed to the fastest old sneak ladder at 7.
I like how current code is continuous climbing instead of jumping, the sneak ladder structures make more sense as climbing structures than as jumping structures. Using the ladders becomes unpleasant when it's a series of jumps, players will have long climbs using sneak ladders so i think smooth climbing is essential.

Concerning your changes to the old move code, this may not be a good idea since it's there to be absolutely identical to old behaviour (although the removal of 'got teleported' may be ok).

I appreciate your work but for various reasons i am :-1: sorry.

'No damage evading' is a non-issue because Fixer has discovered that was a false bug.
'No weird behaviour on stairs': the PR has a new type of weird behaviour, floating over non-existant parts of stairs as if they are full nodes.

**Comment by SmallJoker** [comment]
> i think smooth climbing is essential

I will try to make this smoother. Hopefully there's an easy fix for this. I'm not willing to add much more code just because the glitch isn't "working" as desired.
> your changes to the old move code, this may not be a good idea

The behaviour is identical. I spent enough time on analyzing the functionality.
> 'No damage evading'

This is in regard to sneak ladders, where damage evading currently is possible. I updated the OP to correct this misunderstanding.
> 'No weird behaviour on stairs'

For the sake of KISS there's the deal of floating or the teleport / weird walking bug on stairs.

File: src/localplayer.cpp (hunk 1 above)
Comment (SmallJoker) [inline src/localplayer.cpp]: There are various solutions for this problem:
- Extend the sneak distance -> players will fall down when releasing the sneak key
- Make the player collision box bigger
- Change the stairs' collisionbox

I'm not keen on any of these and would rather keep (halfway) the old sneak behaviour for this situation.

**Comment by paramat** [comment]
Thanks, and sorry for my misunderstanding about damage avoidance.

**Comment by SmallJoker** [comment]
Okay, closer than now is quite hard to reach:

2-row sneak ladder, 90 nodes upwards:
Old code: 14.41s (6.25 n/s), from previous measurements
Latest commit: 14.26s (6.21 n/s)

It also doesn't feel like jumping anymore, it's rather climbing as desired. Opinions?

**Comment by paramat** [comment]
Will test.

**Comment by paramat** [comment]
Warnings on compile:
```In file included from /home//minetest/src/localplayer.cpp:20:0:
/home//minetest/src/localplayer.h: In constructor ‘LocalPlayer::LocalPlayer(Client*, const char*)’:
/home//minetest/src/localplayer.h:157:6: warning: ‘LocalPlayer::m_sneak_node_bb_ymax’ will be initialized after [-Wreorder]
  f32 m_sneak_node_bb_ymax;
      ^
/home//minetest/src/localplayer.h:148:9: warning:   ‘aabb3f LocalPlayer::m_sneak_node_bb_top’ [-Wreorder]
  aabb3f m_sneak_node_bb_top;
         ^
/home//minetest/src/localplayer.cpp:34:1: warning:   when initialized here [-Wreorder]
 LocalPlayer::LocalPlayer(Client *client, const char *name):
 ^
```

**Comment by paramat** [comment]
Testing latest.

2 column sneak ladders are smoother but still with a little jerkiness in the motion.

1 column sneak ladders are unchanged, slower at 5n/s and a series of individual jumps in stead of a smooth motion. Players use single column ladders a lot so these are important. I feel it is imporatant that our new code replacements are at least as fast as old code. Current code has 1 and 2 column ladders both at a smooth 7 n/s which is a little faster than old code.

2 node sneak jump doesn't work at all now.

The first 2 points are enough for me to :-1: Sorry, i feel bad disapproving after all your work.
Maybe you could try applying what you have learnt to current code to attempt fixing the sneaking around ledge edges, damage evasion and jumping at sneak limit, while keeping current sneak ladder and sneak jump behaviour (which is currently very good)?

The idea and priority of current sneak glitch behaviour is to provide good replications of old sneak side-effects, we currently have good behaviour and this PR is a regression.
The enabling of jumping at sneak limit by this PR is not an advantage because in testing it has a very much reduced forward motion in the jump.
The damage evasion and ability to sneak around a ledge corner are good aspects of this PR.

**Comment by SmallJoker** [comment]
@paramat I can not improve the sneak behaviour without changing anything. Deal with the side effects.

Update 170507:
- 2-row sneak ladder speed ≈  jump speed (6.5 n/s)
- Smooth movement on 2-row sneak ladders
- Weird sinking added back in, due smooth 2-row sneak ladders
- *ledge-jumps* as there are no nodes around, Sam has to jump and climb to the next ledge
- Fewer node checks. Worst case scenario: 43 (current), 22 (this PR). Condition: `physics_override_sneak_glitch = true`

Please decide what you want or shoot this PR down if there are no other opinions.

**Comment by paramat** [comment]
Ok will test.

**Comment by paramat** [comment]
Perhaps i have had my priorities slightly wrong. This PR improves general sneak behaviour and surely that must have priority over preserving an old unintended side-effect (sneak ladders). 2-column sneak ladders still work well so that is possibly good enough.

**Comment by paramat** [comment]
Wow 1-column sneak ladder is fast and smooth now too, was not expecting that.

**Comment by paramat** [comment]
Thanks is it ready to retest?

**Comment by SmallJoker** [comment]
@paramat I have no clue how 1/32 slabs work better for you in the current sneak code rather than the previous commits of this pull. Actually it make the slab detection worse now by reverting to the old code.
Whatever, yes. Please retest.

**Comment by paramat** [comment]
> Either there will be abrupt jumps or more slow sinking, I decided for abrupt jumps instead.

Hmm, i consider the old slow sinking on a sneak ladder to be less of a problem than abrupt teleports (if 'teleports' is what you mean by 'jumps').

Will retest 1/32 slabs in master and in this PR again to be sure.

**Comment by paramat** [comment]
> Sneaking fails on a 1/32 nodebox slab (at base of a node) if nothing is underneath it (rare situation though), but if you sneak-jump first it does work. No problem with 1/16th nodebox slab though.
Sneaking fails on a 1/32 nodebox slab placed on a full node, but if you sneak-jump first it does work.
Both these work in master.

Still happening with sneak glitch off

**Comment by paramat** [comment]
Note that the amount of node checks in current code is roughly 8, my sneak-jump code actually enables sneak ladders too, the following checks are just a backup.

**Comment by SmallJoker** [comment]
@paramat strange enough that you get two different results while actually using the same detection algorithm.
Here's the stairs function that I modified to get 1/32 slabs to test the PR:
```Lua
function stairs.register_slab(...)
...
			fixed = {-0.5, -0.5, -0.5, 0.5, 15/16 * -0.5, 0.5},
```
I will analyze this strange issue further in the next days.
> Note that the amount of node checks in current code is roughly 8

Please note the worst-case scenario when checking all 9 nodes + sneak ladder and the ledge detection aswell, that's how I got to these worse-case scenario values.

**Comment by paramat** [comment]
I tested with
`fixed = {-0.5, -0.5, -0.5, 0.5, -15 / 32, 0.5},` as a fraction is more likely used in a nodebox definition.
When i wrote that 1-column sneak ladder is 'smoother' in https://github.com/minetest/minetest/pull/5533#issuecomment-300953655 it was not completely smooth, but it was not jump-and-fall jump-and-fall ... like the most recent version i tested.
So yes something is odd, happy to retest versions in future.

**Comment by paramat** [comment]
Probably too late for this release, what we have currently is fine for now.

**Comment by paramat** [comment]
Ok will test again sometime.

**Comment by paramat** [comment]
> Sneaking fails on a 1/32 nodebox slab (at base of a node) if nothing is underneath it (rare situation though), but if you sneak-jump first it does work. No problem with 1/16th nodebox slab though.
Sneaking fails on a 1/32 nodebox slab placed on a full node, but if you sneak-jump first it does work.
Both these work in master.

Still happening with sneak glitch on or off. However 1/32 nodes are rare so i don't feel this is a big issue, fine for me.

2-node sneak jump works correctly according to 'sneak glitch'.

All sneak ladders are fast and fairly smooth.

Ledge jumps still have a very small forward speed but this doesn't bother me.

2 node sneak-jump still has a abrupt teleport near the top but not a big issue, fine for me.

So all fine for behaviour.

File: src/localplayer.cpp [hunk 14]
```diff
@@ -42,88 +42,134 @@ LocalPlayer::~LocalPlayer()
...
+		if (!is_valid_position || !nodemgr->get(node).walkable)
+			continue;
+		// And the node(s) above have to be nonwalkable
+		bool ok = true;
+		if (!physics_override_sneak_glitch) {
+			u16 height = ceilf(
+					(m_collisionbox.MaxEdge.Y - m_collisionbox.MinEdge.Y) / BS
+			);
+			for (u16 y = 1; y <= height; y++) {
+				node = map->getNodeNoEx(p + v3s16(0,y,0), &is_valid_position);
```
Comment (paramat) [inline src/localplayer.cpp]: Spaces after commas.

File: src/localplayer.cpp [hunk 15]
```diff
@@ -42,88 +42,134 @@ LocalPlayer::~LocalPlayer()
...
+				if (!is_valid_position || nodemgr->get(node).walkable) {
+					ok = false;
+					break;
+				}
+			}
+		} else {
+			// legacy behaviour: check just one node
+			node = map->getNodeNoEx(p + v3s16(0,1,0), &is_valid_position);
```
Comment (paramat) [inline src/localplayer.cpp]: Spaces after commas.

File: src/localplayer.cpp [hunk 16]
```diff
@@ -42,88 +42,134 @@ LocalPlayer::~LocalPlayer()
...
+	node = map->getNodeNoEx(m_sneak_node);
+	std::vector<aabb3f> nodeboxes;
+	node.getCollisionBoxes(nodemgr, &nodeboxes);
+	m_sneak_node_bb_top = getNodeBoundingBox(nodeboxes);
 
-		node = GETNODE(map, pos, vec, 1, &is_valid_position);
+	if (physics_override_sneak_glitch) {
+		// Detect sneak ladder:
+		// Node two meters above sneak node must be solid
+		node = map->getNodeNoEx(m_sneak_node + v3s16(0,2,0),
```
Comment (paramat) [inline src/localplayer.cpp]: Spaces after commas.

File: src/localplayer.cpp [hunk 17]
```diff
@@ -42,88 +42,134 @@ LocalPlayer::~LocalPlayer()
...
+			&is_valid_position);
 		if (is_valid_position && nodemgr->get(node).walkable) {
-			// Ledge exists
-			node = GETNODE(map, pos, vec, 2, &is_valid_position);
-			if (is_valid_position && !nodemgr->get(node).walkable)
-				// Space above ledge exists
-				return true;
+			// Node three meters above: must be non-solid
+			node = map->getNodeNoEx(m_sneak_node + v3s16(0,3,0),
```
Comment (paramat) [inline src/localplayer.cpp]: Spaces after commas.

**Comment by paramat** [comment]
Minor code style stuff but :+1: 
As i don't fully understand the implementation i would like sfan5 to review the code before a merge.

**Comment by paramat** [comment]
@SmallJoker shall we merge this or are you going to attempt to fix the bugs of old code while keeping exact old behaviour? If you do attempt this note that sneak ladder and sneak-jump need to be optional (and disabled by default) which may be impossible.

**Comment by paramat** [comment]
Actually, you can be trusted on this code you know it better than anyone, this can be merged now.

**Comment by paramat** [comment]
Looking at @sfan5 's line comments above in the thread some may be unaddressed, so i think he should review before a merge.

**Comment by paramat** [comment]
Does this fix #5970 ?

**Comment by SmallJoker** [comment]
@paramat Yes. The player's position is no longer modified without a speed reset, which means this bug will not occur here.

**Comment by paramat** [comment]
Ok :+1: 
This fixes that bug, enables sneaking around the corner of a groove in a wall.
1/32nd slabs don't work but are rarely used.

File: src/localplayer.cpp (hunk 1 above)
Comment (paramat) [inline src/localplayer.cpp]: I'm not keen on any of those 3 suggestions either.

**Comment by paramat** [comment]
> Also, if there's no need for a better sneak code (as some people prefer the old one), I don't see a point in working on it.

There is need for improved new sneak, despite some preferring the old one. It would be good to have 1/32 slabs working if possible, but they are rare so it's not high priority.

**Comment by paramat** [comment]
a5c37717ffea387f8aa694f22ccc2f2f449826ac
Also corrected code style.

**Comment by SmallJoker** [comment]
@paramat How about sfan5's review?

**Comment by sfan5** [comment]
I haven't had time to review this again, but since I have reviewed it once before it's probably fine.

**Comment by paramat** [comment]
> sfan5's comments were either answered or solved in a newer commits.

So i changed my mind, i did not see it as essential that he review. I trust your code so i +1 and thought we might as well merge. Hope this is ok.

**Comment by SmallJoker** [comment]
@paramat Another review would've been optimal, as we've both worked on this code. According to the comments I can see that you spent some time on testing the situations, so I don't see a problem there if you accept the current state.
